File: 31-NextPermutation/31-NextPermutation.java
// Last updated: 7/31/2025, 5:38:25 PM
class Solution {
    public void nextPermutation(int[] nums) {
        int p=-1;
        int n=nums.length;
        for(int i=nums.length-2; i>=0; i--){
            if(nums[i]<nums[i+1]){
                p=i;
                break;
            }
        }
        if(p==-1){
            reverse(nums,0,nums.length);
            return;
        }
        else{
            int q=-1;
            for(int i=n-1; i>p; i--){
                if(nums[i]>nums[p]){
                    q=i;
                    break;
                }
            }
            int temp =nums[p];
            nums[p] = nums[q];
            nums[q]= temp;
            reverse(nums, p+1, n);
        }
    }
    public static void reverse(int[] nums, int i, int e){
        while(i<e){
            int te=nums[i];
            nums[i]=nums[e-1];
            nums[e-1]=te;
            i++;
            e--;
        }
    }
    
}

File: 31-NextPermutation/31-NextPermutation.py
# Last updated: 7/31/2025, 5:38:31 PM
from typing import List

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        n = len(nums)
        if n <= 1:
            return
        
        # Step 1: Find the rightmost ascending order pair
        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        
        if i >= 0:  # There is a valid ascending pair
            # Step 2: Find the element to swap with
            j = n - 1
            while nums[j] <= nums[i]:
                j -= 1
            
            # Step 3: Swap elements
            nums[i], nums[j] = nums[j], nums[i]
        
        # Step 4: Reverse the elements after the ith position
        self.reverse(nums, i + 1, n - 1)
    
    def reverse(self, nums: List[int], start: int, end: int) -> None:
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1
